server/sonar-server/src/main/java/org/sonar/server/qualitygate/QgateProjectFinder.java

```diff
@@ -19,22 +19,92 @@
+import static org.sonar.api.utils.Paging.forPageIndex;
 
 @ServerSide
 public class QgateProjectFinder {
 
+  private final DbClient dbClient;
+  private final QualityGateDao qualitygateDao;
+  private final ProjectQgateAssociationDao associationDao;
+  private final UserSession userSession;
+
+  public QgateProjectFinder(DbClient dbClient, UserSession userSession) {
+    this.dbClient = dbClient;
+    this.userSession = userSession;
+    this.qualitygateDao = dbClient.qualityGateDao();
+    this.associationDao = dbClient.projectQgateAssociationDao();
+  }
+
+  public Association find(ProjectQgateAssociationQuery query) {
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      validateId(dbSession, query.gateId());
+      List<ProjectQgateAssociationDto> projects = associationDao.selectProjects(dbSession, query);
+      List<ProjectQgateAssociationDto> authorizedProjects = keepAuthorizedProjects(dbSession, projects);
+
+      Paging paging = forPageIndex(query.pageIndex()).withPageSize(query.pageSize()).andTotal(authorizedProjects.size());
+      return new Association(toProjectAssociations(getPaginatedProjects(authorizedProjects, paging)), paging.hasNextPage());
+    } finally {
+      dbClient.closeSession(dbSession);
+    }
+  }
+
+  private Long validateId(DbSession dbSession, String gateId) {
+    QualityGateDto qualityGateDto = qualitygateDao.selectById(dbSession, Long.valueOf(gateId));
+    if (qualityGateDto == null) {
+      throw new NotFoundException("Quality gate '" + gateId + "' does not exists.");
+    }
+    return qualityGateDto.getId();
+  }
+
+  private static List<ProjectQgateAssociationDto> getPaginatedProjects(List<ProjectQgateAssociationDto> projects, Paging paging) {
+    List<ProjectQgateAssociationDto> pagedAuthorizedDtos = Lists.newArrayList(projects);
+    if (pagedAuthorizedDtos.size() <= paging.offset()) {
+      pagedAuthorizedDtos = Lists.newArrayList();
+    } else if (pagedAuthorizedDtos.size() > paging.pageSize()) {
+      int endIndex = Math.min(paging.offset() + paging.pageSize(), pagedAuthorizedDtos.size());
+      pagedAuthorizedDtos = pagedAuthorizedDtos.subList(paging.offset(), endIndex);
+    }
+    return pagedAuthorizedDtos;
+  }
+
+  private static List<ProjectQgateAssociation> toProjectAssociations(List<ProjectQgateAssociationDto> dtos) {
+    List<ProjectQgateAssociation> groups = newArrayList();
+    for (ProjectQgateAssociationDto groupMembershipDto : dtos) {
+      groups.add(groupMembershipDto.toQgateAssociation());
+    }
+    return groups;
+  }
+
+  private List<ProjectQgateAssociationDto> keepAuthorizedProjects(DbSession dbSession, List<ProjectQgateAssociationDto> projects) {
+    List<Long> projectIds = from(projects).transform(ToProjectId.INSTANCE).toList();
+    Collection<Long> authorizedProjectIds = dbClient.authorizationDao().keepAuthorizedProjectIds(dbSession, projectIds, userSession.getUserId(), UserRole.USER);
```

Comment:
Good question. I think it's because it take a list of project ids in parameter, and it returns a filtered list of projects.